File: apps/web/src/types/message.ts
/**
 * File purpose: Defines shared TypeScript contracts for message.
 * Main declarations: this file contains declarative configuration or re-exports and has no
 * callable declarations.
 */

export type MessageRole = "user" | "assistant" | "system" | "tool";

export interface Message {
  id: string;
  sessionId: string;
  role: MessageRole;
  content: string;
  createdAt: string;
  artifactIds?: string[];
  isPending?: boolean;
  pendingLabel?: string;
  waitStartedAt?: string;
}

File: services/api/app/schemas/message.py
from typing import Literal

from app.schemas.base import ApiModel
from app.schemas.session import Session

MessageRole = Literal["user", "assistant", "system", "tool"]


class Message(ApiModel):
    id: str
    session_id: str
    role: MessageRole
    content: str
    created_at: str
    artifact_ids: list[str] | None = None


class MessageCreate(ApiModel):
    content: str
    model_id: str | None = None


class SendMessageResult(ApiModel):
    messages: list[Message]
    run_id: str | None = None
    session: Session
